db/wal_version.h

```diff
@@ -0,0 +1,129 @@
+
+namespace ROCKSDB_NAMESPACE {
+
+class JSONWriter;
+class Slice;
+class Status;
+
+using WalNumber = uint64_t;
+
+// Metadata of a WAL.
+class WalMetadata {
+ public:
+  WalMetadata() = default;
+
+  explicit WalMetadata(uint64_t size_bytes) : size_bytes_(size_bytes) {}
+
+  bool HasSize() const { return size_bytes_ != kUnknownWalSize; }
+
+  void SetSizeInBytes(uint64_t bytes) { size_bytes_ = bytes; }
+
+  uint64_t GetSizeInBytes() const { return size_bytes_; }
+
+ private:
+  // The size of WAL is unknown, used when the WAL is not closed yet.
+  constexpr static uint64_t kUnknownWalSize = 0;
+
+  // Size of a closed WAL in bytes.
+  uint64_t size_bytes_ = kUnknownWalSize;
+};
+
+// Records the event of adding a WAL in VersionEdit.
+class WalAddition {
+ public:
+  WalAddition() : number_(0), metadata_() {}
+
+  explicit WalAddition(WalNumber number) : number_(number), metadata_() {}
+
+  explicit WalAddition(WalNumber number, const WalMetadata& meta)
+      : number_(number), metadata_(meta) {}
+
+  WalNumber GetLogNumber() const { return number_; }
+
+  const WalMetadata& GetMetadata() const { return metadata_; }
+
+  void EncodeTo(std::string* dst) const;
+
+  Status DecodeFrom(Slice* src);
+
+  std::string DebugString() const;
+
+ private:
+  WalNumber number_;
+  WalMetadata metadata_;
+};
+
+std::ostream& operator<<(std::ostream& os, const WalAddition& wal);
+JSONWriter& operator<<(JSONWriter& jw, const WalAddition& wal);
+
+using WalAdditions = std::vector<WalAddition>;
+
+// Records the event of deleting/archiving a WAL in VersionEdit.
+class WalDeletion {
+ public:
+  WalDeletion() : number_(0) {}
+
+  explicit WalDeletion(WalNumber number) : number_(number) {}
+
+  WalNumber GetLogNumber() const { return number_; }
+
+  void EncodeTo(std::string* dst) const;
+
+  Status DecodeFrom(Slice* src);
+
+  std::string DebugString() const;
+
+ private:
+  WalNumber number_;
+};
+
+std::ostream& operator<<(std::ostream& os, const WalDeletion& wal);
+JSONWriter& operator<<(JSONWriter& jw, const WalDeletion& wal);
+
+using WalDeletions = std::vector<WalDeletion>;
+
+// Used in VersionSet to keep the current set of WALs.
+//
+// When a WAL is created, closed, deleted, or archived,
+// a VersionEdit is logged to MANIFEST and
+// the WAL is added to or deleted from WalSet.
+//
+// Not thread safe, needs external synchronization such as holding DB mutex.
+class WalSet {
+ public:
+  // Add WAL(s).
+  // Can happen when applying a VersionEdit or recovering from MANIFEST.
+  void AddWal(const WalAddition& wal);
+  void AddWals(const WalAdditions& wals);
+
+  // Delete WAL(s).
+  // Can happen when applying a VersionEdit or recovering from MANIFEST.
+  void DeleteWal(const WalDeletion& wal);
+  void DeleteWals(const WalDeletions& wals);
+
+  // Resets the internal state.
+  void Reset();
+
+  const std::map<WalNumber, WalMetadata>& GetWals() const { return wals_; }
+
+ private:
+  std::map<WalNumber, WalMetadata> wals_;
```

Comment:
We could consider using `unordered_map` unless we need this to be sorted.

Reply:
need this to be sorted so that we can check the WALs in increasing order of log number.

Reply:
During recovery check, we need to iterate through `wals_` in increasing order of log number.

Reply:
going to make this unordered_map, then sort the log numbers during recovery.

Reply:
also want to write the logs in increasing order when creating manifest snapshot.